### PR TITLE
feat: API & Domain Association

### LINKED ISSUE #23: Looking to add support for specify a domain for AppSync
Hi there. It looks like AWS recently added support for custom domains on your AppSync APIs: https://aws.amazon.com/blogs/mobile/introducing-custom-domain-names-for-aws-appsync-apis/

It would be great if those module supported that configuration.

Thanks!

### PR DESCRIPTION
**Pull request by svenlito**
closes https://github.com/terraform-aws-modules/terraform-aws-appsync/issues/23

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -88,6 +88,21 @@ resource "aws_appsync_graphql_api" "this" {
   tags = merge({ Name = var.name }, var.graphql_api_tags)
 }
 
+# API Association & Domain Name
+resource "aws_appsync_domain_name" "this" {
+  count = var.create_graphql_api && var.domain_name_association_enabled ? 1 : 0
+
+  domain_name     = var.domain_name
+  certificate_arn = var.certificate_arn
```
Comment (antonbabenko) [inline main.tf:97]: Missing `description` argument.

File: outputs.tf [hunk 2]
```diff
@@ -53,6 +53,22 @@ output "appsync_function_function_id" {
   value       = { for k, v in aws_appsync_function.this : k => v.function_id }
 }
 
+# Domain
+output "appsync_domain_id" {
+  description = "The Appsync Domain Name."
+  value       = var.domain_name_association_enabled ? aws_appsync_domain_name.this[0].id : null
```
Comment (antonbabenko) [inline outputs.tf]:
```suggestion
  value       = try(aws_appsync_domain_name.this[0].id, null)
```
This is better. Please apply a similar fix on other outputs, too. Including what was there before (e.g. `appsync_graphql_api_id` and others).

In some cases, a ternary operation can be broken when resource creation failed but some argument is already being requested. `try` is more optimistic in this case.

File: variables.tf [hunk 3]
```diff
@@ -106,13 +112,27 @@ variable "tags" {
   default     = {}
 }
 
+# API Association & Domain Name
+variable "domain_name" {
+  description = "value"
```
Comment (antonbabenko) [inline variables.tf]: Update `description`.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
   - repo: https://github.com/antonbabenko/pre-commit-terraform
-    rev: v1.62.3
+    rev: v1.64.0
     hooks:
       - id: terraform_fmt
       - id: terraform_validate
       - id: terraform_docs
         args:
-          - '--args=--lockfile=false'
+            - '--args=--lockfile=false'
```
Comment (antonbabenko) [inline .pre-commit-config.yaml]:
```suggestion
          - '--args=--lockfile=false'
```
2 spaces :)

**Review by antonbabenko — CHANGES_REQUESTED** [review]

File: outputs.tf (hunk 2 above)
Comment (svenlito) [reply outputs.tf]: yeah, I don't know what I was thinking. you're absolutely right. Thanks for the feedback :)

File: variables.tf [hunk 5]
```diff
@@ -112,6 +117,25 @@ variable "tags" {
...
+variable "domain_name" {
+  description = "The domain name that AppSync gets associated with."
+  type        = string
+  default     = ""
+}
+
+variable "domain_name_description" {
+  description = "A description of the Domain Name."
+  type        = string
+  default     = ""
```
Comment (antonbabenko) [inline variables.tf]:
```suggestion
  default     = null
```
Since it is optional

**Comment by antonbabenko** [comment]
Examples worked as they should. Merging now. Thank you!

**Comment by antonbabenko** [comment]
This PR is included in version 1.4.0 :tada:

**Pull request merged**